src/Microsoft.AspNet.Routing/RouteData.cs

```diff
@@ -32,24 +34,55 @@ public RouteData(RouteData other)
                 throw new ArgumentNullException(nameof(other));
             }
 
-            DataTokens = new Dictionary<string, object>(other.DataTokens, StringComparer.OrdinalIgnoreCase);
             Routers = new List<IRouter>(other.Routers);
-            Values = new RouteValueDictionary(other.Values);
+            
+            // Perf: Avoid allocating DataTokens and RouteValues unless we need to make a copy.
+            if (other._dataTokens != null)
+            {
+                _dataTokens = new Dictionary<string, object>(other.DataTokens, StringComparer.OrdinalIgnoreCase);
```

Comment:
I'm going to add another change here and measure. The one place the framework interacts with data tokens is in template route. I'm going to make it so that if you don't set data tokens then none of these dictionaries are ever created.

As for the route-values, I don't think copy on write will help too much, because they always get written by template route.